### PR TITLE
Add "volume" parameter to command line and handle reconnection to the agent

### PR DESCRIPTION
**Pull request by sboeuf**
- Add volume parameter to command line
- Fix comment formatting issues
- Check connection status every time start() is called

### CODE REVIEW

File: virtc/main.go [hunk 1]
```diff
@@ -100,6 +100,18 @@ var podConfigFlags = []cli.Flag{
...
+		Name:  "hyper-mount-tag-list",
+		Value: new(cli.StringSlice),
+		Usage: "the hyperstart mount tag list",
+	},
+
+	cli.StringSliceFlag{
+		Name:  "hyper-host-path-list",
+		Value: new(cli.StringSlice),
+		Usage: "the hyperstart host path list",
+	},
```
Comment (sboeuf) [inline virtc/main.go]: I have updated the commit. It now expects an entry like this: -volume="mounttag1:hostpath1 mounttag2:hostpath2 mounttag3:hostpath3"

File: virtc/main.go [hunk 2]
```diff
@@ -100,6 +100,12 @@ var podConfigFlags = []cli.Flag{
 		Value: "",
 		Usage: "the hyperstart tty socket type",
 	},
+
+	cli.StringSliceFlag{
+		Name:  "volume",
+		Value: new(cli.StringSlice),
+		Usage: "the volume to be shared with VM",
+	},
```
Comment (sameo) [inline virtc/main.go]: Here you should declare that one as a `cli.GenericFlag` and then get it directly as a `Volume`. See how virtc/main.go does that for `agentType` for example.

File: agent.go [hunk 3]
```diff
@@ -127,12 +127,15 @@ type agent interface {
 	// start will start the agent on the host.
 	start() error
 
+	// isStarted will check the agent is started.
+	isStarted() bool
```
Comment (sameo) [inline agent.go]: I'd prefer that one to be more generic: `verifyState() agentStateString`. And we should update and potentially store the new agent state based as part of that function.

File: agent.go [hunk 4]
```diff
@@ -127,12 +127,15 @@ type agent interface {
...
+
 	// exec will tell the agent to run a command in an already running container.
 	exec(podID string, contID string, cmd Cmd) error
 
-	// StartPod will tell the agent to start all containers related to the Pod.
+	// startPod will tell the agent to start all containers related to the Pod.
 	startPod(config PodConfig) error
 
-	// StopPod will tell the agent to stop all containers related to the Pod.
+	// stopPod will tell the agent to stop all containers related to the Pod.
```
Comment (sameo) [inline agent.go]: Those 2 are separate changes, so I'd appreciate if you could put that into a separate patch as well.

File: pod.go [hunk 5]
```diff
@@ -680,6 +681,22 @@ func (p *Pod) delete() error {
 	return nil
 }
 
+func (p *Pod) startAgent() error {
+	if p.state.AgentState != agentStarted {
```
Comment (sameo) [inline pod.go]: A more idiomatic code style:
```
if p.state.AgentState == agentStarted {
       return nil
}

err := p.agent.start()
[...]
```

File: hyperstart.go [hunk 6]
```diff
@@ -361,6 +361,16 @@ func (h *hyper) start() error {
 	return nil
 }
 
+// isStarted is the agent state checking implementation for hyperstart.
+func (h *hyper) isStarted() bool {
```
Comment (sameo) [inline hyperstart.go]: We probably want to return `agentStopped` if the agent state is stopped.

File: pod.go [hunk 7]
```diff
@@ -682,7 +682,7 @@ func (p *Pod) delete() error {
 }
 
 func (p *Pod) startAgent() error {
-	if p.state.AgentState != agentStarted {
+	if p.state.AgentState != agentStarted || p.agent.isStarted() == false {
```
Comment (sameo) [inline pod.go]: Could we just call `if p.agent.verifyState() != agentStarted {` here ?

**Comment by sameo** [comment]
@sboeuf Please provide a more descriptive pull request title.

File: pod.go [hunk 8]
```diff
@@ -135,10 +136,50 @@ func (state *PodState) validTransition(oldState stateString, newState stateStrin
 // defined by its mount tag and its host path.
 type Volume struct {
 	// MountTag is a label used as a hint to the guest.
-	MountTag string
+	mountTag string
 
 	// HostPath is the host filesystem path for this volume.
-	HostPath string
+	hostPath string
```
Comment (sameo) [inline pod.go]: We need to keep those exported as it is part of our `PodConfig` structure.

File: pod.go [hunk 9]
```diff
@@ -135,10 +136,50 @@ func (state *PodState) validTransition(oldState stateString, newState stateStrin
...
+}
+
+// Volumes is a Volume list.
+type Volumes []Volume
+
+// Set assigns volume values from string to a Volume.
+func (v *Volumes) Set(volStr string) error {
+	volArray := strings.Split(volStr, " ")
```
Comment (sameo) [inline pod.go]: Nitpicking here: `volSlice` would be a more accurate name.

File: hyperstart.go [hunk 10]
```diff
@@ -315,6 +347,38 @@ func buildHyperContainerProcess(cmd Cmd) (hyperJson.Process, error) {
 	return process, nil
 }
 
+func isStarted(c net.Conn, chType chType) bool {
+	ret := false
+	timeoutDuration := 5 * time.Second
```
Comment (sameo) [inline hyperstart.go]: 5s is too much imho. If we can't ping in less than a second, we're not connected.

File: hyperstart.go [hunk 11]
```diff
@@ -400,10 +468,17 @@ func (h *hyper) startPod(config PodConfig) error {
 		containers = append(containers, container)
 	}
 
+	var mountTag string
+	if h.config.Volumes != nil {
+		mountTag = h.config.Volumes[0].mountTag
+	} else {
+		mountTag = ""
+	}
+
```
Comment (sameo) [inline hyperstart.go]: Not sure this chunk is related to this patch ?

File: hyperstart.go [hunk 12]
```diff
@@ -315,6 +347,38 @@ func buildHyperContainerProcess(cmd Cmd) (hyperJson.Process, error) {
 	return process, nil
 }
 
+func isStarted(c net.Conn, chType chType) bool {
```
Comment (sameo) [inline hyperstart.go]: We probably only need to check for one channel only. I can't see a case where we'd have only one channel alive and not the other.
